N-dimensional interpolation

kylecarow opened the pull request:
# Issue
https://github.nrel.gov/MBAP/fastsim/issues/328

# Completed 
- [x] investigate dimensionality reduction
  - removing dimensionality reduction for hardcoded stuff sped it up, but multilinear still benefits from it, probably due to the 'views' that ndarray employs for this
- [x] investigate `.windows()` vs. binary search for finding indices
  - on a benchmark of 100x100x100 3d data (using InterpND) the binary search from RouteE and the `.windows` search perform almost exactly the same, so I'll keep it as is. This is easy to test by placing the index finding code into a function and swapping out the function body

# Fairly Urgent
- [x] get rid of standalone `interp1d` everywhere and replace with `Interpolator`

# Not Urgent but Important
- [ ] put `PhantomData` in pretty much every named field struct and maybe also the tuple structs to prevent `init` bypassing
- [x] manually generate jsonized getters and setters for `Interpolator` and maybe other enums
- [x] automatically generate jsonized getters and setters for `Interpolator` and maybe other enums, making use of [`FieldOptions`](https://github.com/NREL/fastsim/blob/7b345b23d69ca638b52b54a635a4ed13b1614131/fastsim-core/fastsim-proc-macros/src/fastsim_api/api_utils.rs#L356)

kylecarow commented:
@calbaker fixed the failing tests here

kylecarow commented:
@nreinicke made some good points:
- the dimensionality lowering code does cloning to create a lower dim interpolator that also owns the data
  - I think this dimensionality lowering stuff could be unnecessary, the performance when passing a grid point value is probably not that bad compared to cloning
- binary search instead of using `.windows` may be faster

EDIT: resolved, see top comment

kylecarow commented:
I want to switch to binary search rather than `.windows` and move the other unwrap rewrites as I did in compass

kylecarow commented:
@calbaker This is ready to review once tests pass. ~I'll add benchmarks for the 1 & 2 dimensional cases later on in `fastsim-3`~ I added these in 1412dc2.

I'd appreciate your review on the structure and usage that I've designed here, along with any other thoughts you have!

kylecarow commented:
@calbaker Maybe we can review this in the FASTSim development meeting today

calbaker commented on fastsim-core/benches/interp_bench.rs:
```diff
@@ -0,0 +1,195 @@
+    }
+}
+
+pub fn criterion_benchmark(c: &mut Criterion) {
```
@kylecarow , could you add a doc string here-ish to explain how to trigger the benchmarks?

kylecarow commented:
Converted to draft so I can implement interpolation throughout FASTSim before merging

kylecarow commented:
@calbaker Uh oh

https://pyo3.rs/v0.21.2/class.html?highlight=enum#complex-enums
> An enum is complex if it has any non-unit (struct or tuple) variants.
>
> Currently PyO3 supports only struct variants in a complex enum. Support for unit and tuple variants is planned.

What I have written is an enum with tuple variants that contain structs (as there is no way to impl on enum struct variants). This will take some thinking to solve.

kylecarow commented:
@robinsteuteville 

- We need ways to interact with the data within the `Interpolator` enum (e.g. `x`, `y`, ... `f(x, y, ...)`) from Python, like setting x-values or something from Python
- PyO3 doesn't support enums with tuple variants, so replacing component efficiencies with the new `Interpolator` enum will not work out of the box
- It can work by doing something like this:
    ```rust
    #[api(skip_get, skip_set)]
    pub eff_interp: utils::interp::Interpolator,
    ```
    But then we can't access the variable at all from Python (because we skip adding the getter and setter with the attribute), and we would not be able to access any of its methods (which is where we would have the stuff to achieve the first bullet above)
  - This is what you see when trying to get via `sd.veh.fc.eff_interp`: `AttributeError: 'fastsim.FuelConverter' object has no attribute 'eff_interp'`
- I made a 'wrapper' struct that can be seen from Python without issue, as is in the latest commit in this PR:
    ```rust
    pub eff_interp: utils::interp::InterpolatorWrapper,
    ```
    Getting this from Python works `sd.veh.fc.eff_interp`: `<InterpolatorWrapper at 0x176dc26b0>`
  - We can make the methods to get x-values etc. on this (probably in a `#[pyo3_api(...)]` block, no need to have Rust-side methods when Rust users can edit the data directly)

We also need to fix the tests that fail (expecting the old struct format)

robinsteuteville committed:
keeping branch up to date with fastsim-3

robinsteuteville committed:
Merge branch 'f3/interpolation' of https://github.com/NREL/fastsim into f3/interpolation

calbaker commented:
- [ ] finish #146 when this is done

Commit:
changed threshold down a little bit

robinsteuteville committed:
adding interpolators to electric_machine.rs (still needs to be added to simdrive.rs and reversible_energy_storage.rs)

Commit:
made use of LazyCell for speedup

Commit:
LazyCell needs to be fixed or replaced with `lazy_static`

robinsteuteville committed:
attempting to fix macros.rs errors resulting from interpolation changes

Commit:
variable name change/improvement

Commit:
clippy hint

Commit:
added code to write cycle keys into dataframe

Commit:
fixed an egregious boolean error!

Commit:
typo fix

Commit:
fixed mutability problems

Commit:
Merge branch 'f3/interp-with-benchmarking' into f3/hev-troubleshooting
an test that should fail is failing
`LazyCell` -> `lazy_static` for actual speedup
fixed mutability problem

Commit:
removed `InterpolatorWrapper`
changed `pyo3_api` to `fastsim_api`

robinsteuteville committed:
attempting to troubleshoot macros.rs errors

Commit:
comments, formatting, and work towards getting Interpolator used

Commit:
Interpolation is way broken but Robin will fix it

Commit:
Merge branch 'f3/hev-troubleshooting' into f3/interpolation

Commit:
reverted to subprocess

robinsteuteville committed:
getting rid of interpolator wrapper

Commit:
Merge branch 'f3/interpolation' of github.com:NREL/fastsim into f3/interpolation

Commit:
lots of errors to unwind

Commit:
`as_ref` wins the day.  working through more errors

Commit:
incremental changes

Commit:
more cleanup

robinsteuteville committed:
fixing errors in macros.rs

robinsteuteville committed:
fixing all errors not related to phantom data

robinsteuteville committed:
making x, f_x, etc. interpolator fields private so that any change to them must include calling validate(). Made interpolator related macros into functions, and updated the macros-turned-functions to include both eff_interp_fwd and eff_interp_bwd.

robinsteuteville committed:
updating interpolator fields to be semi-public

robinsteuteville committed:
fixing some of the failed tests and getting rid of to_owned() wherever possible

Commit:
Merge branch 'fastsim-3' of github.com:NREL/fastsim into f3/interpolation
HEV fails but other tests pass

Commit:
Merge branch 'fastsim-3' into f3/interpolation

Commit:
`#[fastsim_api(...)]` needs to come before `#[derive(Deserialize, Serialize...)]`

Commit:
in progress toward getting cycle stuff but not other stuff all handled in history

Commit:
got `to_dataframe` working and updated plotting to use it

Commit:
moved logging init inside explicit function

Commit:
having `pyo3_log_init` as a user-called function
seems to allow for logging capability without any performance penalty unless it is used

Commit:
fixed power limit problem by setting fc limits in HEV
changed to using si ZERO

kylecarow committed:
ArrayBase with Dyn dim changed to ArrayD

kylecarow committed:
move 1/2/3/N-d code into respective mods

kylecarow committed:
serde(default) rather than serde(skip) for extrapolate field

Commit:
Merge branch 'f3/interpolation' of github.com:NREL/fastsim into f3/interpolation

Commit:
HEV limits and controls need a thorough inspection

Commit:
changed `as_mut` to `as_ref` where feasible

Commit:
powers were not being limited correctly
but now the problem is aux load should be transferred to engine or something else when SOC gets too small to support

Commit:
all rust tests pass. working on pytests

Commit:
all tests pass. ready to merge
